cpp/src/arrow/memory_pool.cc

```diff
@@ -44,104 +48,175 @@ namespace {
 #else
-    std::free(ptr);
+      free(ptr);
 #endif
+    }
   }
-}
+};
+
+#ifdef ARROW_JEMALLOC
 
-Status ReallocateAligned(int64_t old_size, int64_t new_size, uint8_t** ptr) {
-  uint8_t* previous_ptr = *ptr;
-  if (previous_ptr == zero_size_area) {
-    DCHECK_EQ(old_size, 0);
-    return AllocateAligned(new_size, ptr);
+// Helper class directing allocations to the jemalloc allocator.
+class JemallocAllocator {
+ public:
+  static Status AllocateAligned(int64_t size, uint8_t** out) {
+    if (size == 0) {
+      *out = zero_size_area;
+      return Status::OK();
+    }
+    *out = reinterpret_cast<uint8_t*>(
+        mallocx(static_cast<size_t>(size), MALLOCX_ALIGN(kAlignment)));
+    if (*out == NULL) {
+      return Status::OutOfMemory("malloc of size ", size, " failed");
+    }
+    return Status::OK();
   }
-  if (new_size == 0) {
-    DeallocateAligned(previous_ptr, old_size);
-    *ptr = zero_size_area;
+
+  static Status ReallocateAligned(int64_t old_size, int64_t new_size, uint8_t** ptr) {
+    uint8_t* previous_ptr = *ptr;
+    if (previous_ptr == zero_size_area) {
+      DCHECK_EQ(old_size, 0);
+      return AllocateAligned(new_size, ptr);
+    }
+    if (new_size == 0) {
+      DeallocateAligned(previous_ptr, old_size);
+      *ptr = zero_size_area;
+      return Status::OK();
+    }
+    *ptr = reinterpret_cast<uint8_t*>(
+        rallocx(*ptr, static_cast<size_t>(new_size), MALLOCX_ALIGN(kAlignment)));
+    if (*ptr == NULL) {
+      *ptr = previous_ptr;
+      return Status::OutOfMemory("realloc of size ", new_size, " failed");
+    }
     return Status::OK();
   }
-#ifdef ARROW_JEMALLOC
-  if (new_size < 0) {
-    return Status::Invalid("negative realloc size");
+
+  static void DeallocateAligned(uint8_t* ptr, int64_t size) {
+    if (ptr == zero_size_area) {
+      DCHECK_EQ(size, 0);
+    } else {
+      dallocx(ptr, MALLOCX_ALIGN(kAlignment));
+    }
   }
-  if (static_cast<uint64_t>(new_size) >= std::numeric_limits<size_t>::max()) {
-    return Status::CapacityError("realloc overflows size_t");
+};
+
+#endif  // defined(ARROW_JEMALLOC)
+
+#ifdef ARROW_MIMALLOC
+
+// Helper class directing allocations to the mimalloc allocator.
+class MimallocAllocator {
+ public:
+  static Status AllocateAligned(int64_t size, uint8_t** out) {
+    if (size == 0) {
+      *out = zero_size_area;
+      return Status::OK();
+    }
+    *out = reinterpret_cast<uint8_t*>(
+        mi_malloc_aligned(static_cast<size_t>(size), kAlignment));
+    if (*out == NULL) {
+      return Status::OutOfMemory("malloc of size ", size, " failed");
+    }
+    return Status::OK();
   }
-  *ptr = reinterpret_cast<uint8_t*>(
-      rallocx(*ptr, static_cast<size_t>(new_size), MALLOCX_ALIGN(kAlignment)));
-  if (*ptr == NULL) {
-    *ptr = previous_ptr;
-    return Status::OutOfMemory("realloc of size ", new_size, " failed");
+
+  static Status ReallocateAligned(int64_t old_size, int64_t new_size, uint8_t** ptr) {
+    uint8_t* previous_ptr = *ptr;
```

Comment:
Can we use `auto` here?

Reply:
I'd rather keep the type explicit for clarity.